refactor(order): orderService 코드 리팩터링

yawning5 opened the pull request:
## OrderService
- 튜터님 피드백 반영하여 메서드들의 코드를 줄였습니다
  - 책임분리 원칙에 따른 메서드 분리
  - 메서드 오버로딩 활용으로 권한 확인
  - 제네릭 타입을 활용한 Repo 조회값 확인 메서드
  - Entity 조회한 뒤 Dto 와 검증 -> 쿼리로 검증

### getOrder, getOrders
- CUSTOMER 권한을 가진 사용자가 로그인된 userId 와 다른 userId 를 가진 데이터 조회시 권한없음 오류를 발생하도록 하였습니다

### createOrder
- 이번에 리뷰주셨던 내용을 바탕으로 존재하는 아이디 조회 검증을 Set 으로 하도록 변경하였습니다

## OrderController
- 권한 설정추가 하였습니다

### createOrder
- CUSTOMER 권한만 수행가능하게 어노테이션 설정

### 그외 메서드
- 로그인한 사용자라면 모두 사용가능합니다
- 서비스로직에서 권한 분기 됩니다

##  확인사항
- 컴파일 오류 발생하지 않는 것을 확인하였습니다
- 테스트 관련 코드는 권한 추가로 인해 이후 Pr 에서 업데이트해 수정하도록 하겠습니다

yawning5 committed:
fix(order): OrderRepository 부분 주석 처리된 코드 아래로 위치이동

주석처리한 코드를 남겨둔 이유를 추가하였습니다

yawning5 committed:
Merge branch 'dev' into refactor/order/orderService

yawning5 committed:
refactor(order):

OrderRequestDto: storeId 를 받아오도록 수정
findPageIds: 검색조건을 Dto 상태로 Repo 에 건네주는 것으로 변경
OrderServiceTest: reqDto 변경으로 인한 테스트 조건 수정
OrderService: 리팩토링을 위해 역할별 코드줄 우선 구분

yawning5 committed:
refactor(order): 튜터님 피드백 적용하여 OrderService.createOrder 를 리팩토링하여 코드 경량화 및 책임 분리 수행

테스트코드 변경사항은 OrderService 리팩토링에 따른 인자변화로 수정하였습니다

yawning5 committed:
feat(order): getOrder 관련기능 사용시 CUSTOMER 권한이면 자기 자신의 주문내역만 조회가능하게 제한

createOrder 컨트롤러는 CUSTOMER 권한만 접근 가능하게 변경

OrderRepositoryCustomImpl PageableExecutionUtils 적용하여 최소쿼리 1회로 변경

기타 주석오타 수정 및 일부 주석 제거 주석 설명 업데이트

컴파일 오류발생 하지 않는것 확인

yawning5 committed:
Merge branch 'dev' into refactor/order/orderService

yawning5 committed:
fix(order): hasPermission 조건문 수정

Doritosch commented on src/main/java/com/sparta/tdd/domain/order/service/OrderService.java:
```diff
@@ -44,106 +44,168 @@ public Page<OrderResponseDto> getOrders(
+            UserDetailsImpl userDetails,
+            OrderSearchOptionDto searchOption) {
+
+        if (userDetails.getUserAuthority() == UserAuthority.CUSTOMER
```
user.enums.UserAuthority에서 `isCustomer()` 메서드를 제공하고 있어서 `UserAuthority.isCustomer(userDetails.getUserAuthority())`처럼 사용하시면 좋을 것 같습니다,.

yawning5 replied:
바로 userDetails 에서 빼와도 괜찮지 않을까요?

Doritosch replied:
<del>제가 생각하기에는 서비스 계층에서 직접 권한을 비교하고 있어, 권한 검증의 책임이 분산된 것 같아 보입니다.</del> 
방식에는 잘못된 것이 없지만 UserAuthority에서 이미 권한 검증 메서드를 제공하고 있으므로 해당 메서드를 재사용한다면 가독성과 일관성을 확보할 수 있을 것 같습니다.

yawning5 replied:
수정하겠습니다!!

Doritosch commented on src/main/java/com/sparta/tdd/domain/order/service/OrderService.java:
```diff
@@ -44,106 +44,168 @@ public Page<OrderResponseDto> getOrders(
+            OrderSearchOptionDto searchOption) {
+
+        if (userDetails.getUserAuthority() == UserAuthority.CUSTOMER
+        && !userDetails.getUserId().equals(searchOption.userId())) {
```
단순히 접근 권한을 검증하는 로직 같은데, 컨트롤러 계층에서 @PreAuthorize로 해결하는 방법은 어떤가요?

yawning5 replied:
저도 그걸 좀 생각해서 적용하려고 해봤는데
Dto 내부의 Id 랑 비교하는 메서드를 사용할때 
어노테이션 적용 클래스에서 따로 메서드를 만들어서 적용하는 방식이더라구요 그래서 메서드로 나누었습니다
혹시 다른 방법이 있다면 알려주시면 감사하겠습니다!

Doritosch replied:
그렇다면 UserAuthority enum에서 동일한 id인지 검증하는 메서드를 추가하거나 현재 사용하는 방식을 사용해도 좋을 것 같습니다.

Doritosch commented on src/main/java/com/sparta/tdd/domain/order/service/OrderService.java:
```diff
@@ -44,106 +44,168 @@ public Page<OrderResponseDto> getOrders(
+     * @param id
+     * @return Entity
+     */
+    private <T, ID> T isPresent(JpaRepository<T, ID> jpaRepository, ID id) {
```
`isPresent()` 메서드 명만 봤을 때는 반환 타입이 boolean일 것이다! 라고 예상했었습니다. 실제로 Entity 객체를 반환하는거라면 `find~()`가 적절하지 않을까 생각합니다.

yawning5 replied:
findEntity 로 수정하였습니다!

Doritosch commented on src/main/java/com/sparta/tdd/domain/order/service/OrderService.java:
```diff
@@ -44,106 +44,168 @@ public Page<OrderResponseDto> getOrders(
-        // 빠진 메뉴(존재하지 않는 menuId) 검증
-        verifyOrderMenus(menuMap, menuIds);
-
         for (OrderMenuRequestDto om : reqDto.menu()) {
-            Menu menu = menuMap.get(om.menuId());
-
-            if (!menu.getStore().getId().equals(foundStore.getId())) {
-                throw new IllegalArgumentException("해당 가게의 메뉴가 아닙니다: menuId=" + om.menuId());
-            }
-
             OrderMenu orderMenu = OrderMenu.builder()
                 .quantity(om.quantity())
                 .price(om.price())
-                .menu(menu)
+                .menu(menuMap.get(om.menuId()))
                 .build();
 
             order.addOrderMenu(orderMenu);
         }
 
-        Order savedOrder = orderRepository.save(order);
-
-        OrderResponseDto resDto = orderMapper.toResponse(savedOrder);
-
-        return resDto;
+        return order;
```
매핑 및 연관관계를 설정하는 책임은 개인적으로 orderService 보다는 entity나 mapper가 가지고 있다고 생각하는데 의현님은 어떻게 생각하시는지 궁금합니다!

yawning5 replied:
저도 동일하게 생각하고 이부분에서 고민을 햇었습니다
이 부분을 Ordrer 나 OrderMapper 에 넘길까 고민했었는데 Menu 엔티티를 다루다 보니까 Order 의 다른 곳으로 넘어가게 되면 Menu 엔티티를 알아야하는 문제가 생겨서 Service 단에서 조합하고 있습니다
OrderMenu 를 만들고 나서는 Order 에서 연관관계 메서드를 불러와서 처리하고 있습니다

Doritosch replied:
order와 menu를 다루는 mapper를 새로 만드는건 어떨까요? 매핑과 연관관계 설정이 서비스 계층에서 이루어지는게 불편하긴 한데 저도 어떻게 해결해야 할지 고민이 되네요.

yawning5 replied:
찾아봤는데 Order 가 Menu 를 바라보는게 매핑만 한다면 크게 문제가 되지는 않는거 같아서
OrderMapper 로 옮기도록 하겠습니다!!
좋은 리뷰 감사드립니다!!

yawning5 replied:
수정 완료하였습니다!!

yawning5 committed:
fix(order):

리뷰적용
UserDetails.getAuthority() 로꺼내와서 비교하는 대신 UserAuthority에서 제공하는 isCustomer 사용하는 것으로 변경

isPresent -> findEntity 로 메서드명 변경

yawning5 committed:
Merge branch 'dev' into refactor/order/orderService

yawning5 committed:
fix(order):

리뷰적용
OrderService 에서 매핑 및 연관관계 설정하던 메서드를 OrderMapper 로 변경

yawning5 committed:
Merge branch 'dev' into refactor/order/orderService

Doritosch reviewed (approved):
고생하셨습니다!
